rust-runtime/aws-smithy-http-server/src/plugin/check_health.rs

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use std::future::Ready;
+
+use futures_util::Future;
+use http::StatusCode;
+use hyper::{Body, Request, Response};
+use tower::Layer;
+use tower::Service;
+
+use crate::body;
+use crate::body::BoxBody;
+
+use super::Either;
+
+#[derive(Clone)]
+pub struct CheckHealthLayer<H> {
+    ping_handler: H,
+}
+
+impl<H> CheckHealthLayer<H> {
+    pub fn new(ping_handler: H) -> Self {
+        CheckHealthLayer { ping_handler }
+    }
+}
+
+pub type DefaultHandler<E> = fn(Request<Body>) -> Ready<Result<Response<BoxBody>, E>>;
+
+impl CheckHealthLayer<()> {
+    pub fn with_default_handler<E>() -> CheckHealthLayer<DefaultHandler<E>> {
+        CheckHealthLayer::new(default_ping_handler)
+    }
+}
+
+impl<S, H: Clone> Layer<S> for CheckHealthLayer<H> {
+    type Service = CheckHealthService<H, S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        CheckHealthService {
+            inner,
+            layer: self.clone(),
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct CheckHealthService<H, S> {
+    inner: S,
+    layer: CheckHealthLayer<H>,
+}
+
+impl<H, HandlerFuture, S> Service<Request<Body>> for CheckHealthService<H, S>
+where
+    S: Service<Request<Body>, Response = Response<BoxBody>> + Clone,
+    S::Future: std::marker::Send + 'static,
+    HandlerFuture: Future<Output = Result<Response<BoxBody>, S::Error>>,
+    H: Fn(Request<Body>) -> HandlerFuture,
+{
+    type Response = S::Response;
+
+    type Error = S::Error;
+
+    type Future = Either<HandlerFuture, S::Future>;
+
+    fn poll_ready(&mut self, cx: &mut std::task::Context<'_>) -> std::task::Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: Request<Body>) -> Self::Future {
+        let clone = self.inner.clone();
```

Comment:
I don't think we need to `clone` here? Can't we just `self.service.call`?

Reply:
This is not the case if we `oneshot` [(b) in this comment](https://github.com/awslabs/smithy-rs/pull/2540/files#r1158283055)

Reply:
Used `oneshot` in bdbed1b012aa191884021398850deefb5a7e63f0.